File: Tutorials/Scala/3 PublishSubscribe/producer/src/main/scala/ru/example/rmq/publishsubscribe/Producer.scala
package ru.example.rmq.publishsubscribe

import com.rabbitmq.client._
import java.nio.charset.StandardCharsets
import scala.util.Using

object Producer {
  def main(args: Array[String]): Unit = {
    val EXCHANGE_NAME = "logs"

    if (args.length < 3) {
      System.out.println("Usage: Producer hostname username password [message]")
      System.exit(-1)
    }

    val hostname    = args(0)
    val username    = args(1)
    val password    = args(2)
    val virtualHost = "/"

    val message =
      if (args.length > 3) args.slice(4, args.length).mkString(" ")
      else "info: Hello World!"

    val factory = new ConnectionFactory
    factory.setHost(hostname)
    factory.setUsername(username)
    factory.setPassword(password)
    factory.setVirtualHost(virtualHost)

    Using.Manager { use =>
      try {
        val connection = use(factory.newConnection)
        val channel    = use(connection.createChannel)

        channel.exchangeDeclare(EXCHANGE_NAME, BuiltinExchangeType.FANOUT)
        channel.basicPublish(EXCHANGE_NAME, "", null, message.getBytes(StandardCharsets.UTF_8))
        println(s"[x] Sent '$message'")
      } catch {
        case e: Exception =>
          println(e.getLocalizedMessage)
          sys.exit(-1)
      }
    }
    sys.exit(0)
  }
}

File: Tutorials/Java/3 PublishSubscribe/Producer/src/main/java/Producer.java
import com.rabbitmq.client.BuiltinExchangeType;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

public class Producer {
    private final static String EXCHANGE_NAME   = "logs";
    private final static String user            = "user";
    private final static String password        = "password";
    private final static String virtualHost     = "/";

    public static void main(String[] argv) throws Exception {
        String host;
        String message;

        if (argv.length > 0) host = argv[0];
        else host = "localhost";

        if (argv.length < 2) message = "info: Hello World!";
        else message = String.join(" ", Arrays.copyOfRange(argv, 1, argv.length));

        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(host);
        factory.setUsername(user);
        factory.setPassword(password);
        factory.setVirtualHost(virtualHost);

        try (Connection connection = factory.newConnection();
             Channel channel = connection.createChannel()) {
            channel.exchangeDeclare(EXCHANGE_NAME, BuiltinExchangeType.FANOUT);
            channel.basicPublish(EXCHANGE_NAME, "", null, message.getBytes(StandardCharsets.UTF_8));
            System.out.println(" [x] Sent '" + message + "'");
        }
    }
}
